c/shared/source/uforead/uforead.c

```diff
@@ -953,174 +956,170 @@ static void setFontDictKey(ufoCtx h, char* keyValue) {
-        char* copySymbol;
-        top->Notice.ptr = keyValue;
-        /* look for the (c) symbol U+00A9, which is 0xC2, 0xA9 in UTF-8 */
-        copySymbol = strstr(keyValue, "\xC2\xA9");
-        if (copySymbol != NULL) {
-            /* if there is a copyright symbol (U+00A9),
-               replace it with the word "Copyright" */
-            char* cpy = "Copyright";
-            char* newString = memNew(h, strlen(cpy) + strlen(keyValue) + 2);
-            /* set the 0xC2 to NULL to terminate the left side of the string */
-            *copySymbol = '\0';
-            /* use copySymbol + 2 to skip the NULL and the 0xA9
-               to get the right side of the string */
-            sprintf(newString, "%s%s%s", keyValue, "Copyright", copySymbol + 2);
-            top->Notice.ptr = newString;
-        }
-    } else if (!strcmp(keyName, "versionMajor")) {
-        if (top->version.ptr == NULL)
-            top->version.ptr = keyValue;
-        else {
-            char* newString = memNew(h, strlen(top->version.ptr) + strlen(keyValue) + 2);
-            sprintf(newString, "%s.%s", keyValue, top->version.ptr);
-            memFree(h, top->version.ptr);
-            top->version.ptr = newString;
-            memFree(h, keyValue);
-        }
-    } else if (!strcmp(keyName, "versionMinor")) {
-        if (top->version.ptr == NULL)
-            top->version.ptr = keyValue;
-        else {
-            char* newString = memNew(h, strlen(top->version.ptr) + strlen(keyValue) + 2);
-            sprintf(newString, "%s.%s", top->version.ptr, keyValue);
-            memFree(h, top->version.ptr);
-            top->version.ptr = newString;
-            memFree(h, keyValue);
-        }
-    } else if (!strcmp(keyName, "postscriptFontName")) {
-        fd->FontName.ptr = keyValue;
-    } else if (!strcmp(keyName, "openTypeNamePreferredFamilyName")) {
-        top->FamilyName.ptr = keyValue;
-    } else if (!strcmp(keyName, "familyName")) {
-        if (top->FamilyName.ptr == NULL)  // we don't re-set this if it was set by "openTypeNamePreferredFamilyName"
+
+    if (!strcmp(keyName, "postscriptFDArray")) {
+        h->top.FDArray.array = memNew(h, FDArrayInitSize *sizeof(abfFontDict));
+        if (h->top.version.ptr != NULL)
+            h->top.cid.CIDFontVersion = atoi(h->top.version.ptr) % 10 + (float) atoi(&h->top.version.ptr[2])/1000;
+        abfInitFontDict(h->top.FDArray.array);
+
+        parsingFDArray = true;
+        currentiFD = -1;
+        parseKeyValue(h, cur);
+        parsingFDArray = false;
+    } else if (!strcmp(keyName, "PrivateDict")) {
+        parsingFDArray = false;
+        parseKeyValue(h, cur);
+        parsingFDArray = true;
+    } else {
+        char* keyValue = (char*) parseKeyValue(h, cur);
+        if (keyValue != NULL && !strcmp(keyValue, ""))
+            message(h, "Warning: Encountered empty <%s> for fontinfo key %s. Skipping", cur->name, keyName);
+        if (!strcmp(keyName, "copyright")) {
```

Comment:
Looks like from a few samples we've only got the additions above and some indentation changes below so I'm reviewing on that basis. Let me know if there are more substantive changes among the indentations that I should look at.

Reply:
There are some general indent changes, but the biggest change for this function is at the top where now it first checks if it is at the beginning of parsing an FDArray. It checks for the postscriptFDArray key, and then does some initialization/etc for FDArray, before checking for any other values. I had to do it this way to make the FDArray parsing process recursive (I should add more documentation about this, so I'll do that).

Reply:
Another big change is the way keyName and keyValue is handled. Previously, keyValue was passed into the function, and keyName was held in h->parseKeyName.